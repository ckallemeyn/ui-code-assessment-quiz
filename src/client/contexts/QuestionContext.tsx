import { createContext } from "react"

// maybe play around with creating an HOC for the entire app
const QuestionContext = createContext({});
QuestionContext.displayName = 'QuestionContext';

export default QuestionContext;
