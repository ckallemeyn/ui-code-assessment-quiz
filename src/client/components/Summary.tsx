import React, { useContext } from 'react'
import { Button, Container, Grid, GridColumn, GridRow } from 'semantic-ui-react'
import { configureAnswers } from '../../utils/questionUtils';
import QuestionContext from '../contexts/QuestionContext'


const SummaryPage = () => {
  const { state, dispatch }: any = useContext(QuestionContext);
  const {
    questions,
    idx,
    correctAnswers,
    incorrectAnswers,
    questionsAnswered,
    finalScorePercentage,
  } = state;

  const restartQuiz = () => {
    // handle if the
    const nextIdx = idx + 1;
    let nextQuestion = questions[nextIdx];
    // basic
    if (nextQuestion === undefined || idx >= questions.length) {
      nextQuestion = questions[0];
    }

    const payload: any = {
      currentQuestion: nextQuestion,
      idx: nextIdx,
      currentAnswer: "",
      correctAnswers: 0,
      incorrectAnswers: 0,
      questionsAnswered: 0,
      finalScorePercentage: 0,
      isSummaryVisible: false,
    };

    if (nextQuestion.incorrect_answers) {
      payload.answers = configureAnswers(nextQuestion);
    }

    return dispatch(({ type: 'RESTART_QUIZ', payload }));
  }

  return (
    <>
      <Container>
        <Grid columns={1}>
          <GridRow>
            <GridColumn>
              <h2>Summary</h2>
              <p>
                Correct: <strong> {correctAnswers} </strong>
              </p>
              <p>
                Incorrect: <strong> {incorrectAnswers} </strong>
              </p>
              <p>
                Total Answered: <strong> {questionsAnswered} </strong>
              </p>
              <p>
                Final Score: <strong> {finalScorePercentage}% </strong>
              </p>
              <Button type="text" primary onClick={restartQuiz}> Restart Quiz </Button>
            </GridColumn>
          </GridRow>
        </Grid>
      </Container>
    </>
  )
}

export default SummaryPage
