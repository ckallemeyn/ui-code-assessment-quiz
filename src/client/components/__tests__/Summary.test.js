import React from "react";
import { render, fireEvent, waitFor } from "@testing-library/react";
import Summary from "../Summary";
import QuestionContext from '../../contexts/QuestionContext'

const renderWithContext = (value) => {
  return render(
    <QuestionContext.Provider value={value}>
      <Summary />
    </QuestionContext.Provider>
  );
}

let contextValue;

beforeEach(() => {
  contextValue = {
    state: {
      correctAnswers: 0,
      incorrectAnswers: 0,
      questionsAnswered: 0,
      finalScorePercentage: 0,
    },
    dispatch: jest.fn()
  }
})



describe("Summary", () => {
  test("it should render", () => {
    const { container } = renderWithContext(contextValue);

    expect(container).toMatchSnapshot();
  });


  test('it should contain a summary of the previous quiz', () => {
    contextValue.state = {
      correctAnswers: 2,
      incorrectAnswers: 3,
      questionsAnswered: 5,
      finalScorePercentage: 40,
    }
    const { getByTestId } = renderWithContext(contextValue);
    const correctAnswers = getByTestId('correct-answers').textContent.trim();
    const incorrectAnswers = getByTestId('incorrect-answers').textContent.trim();
    const questionsAnswered = getByTestId('total-answered').textContent.trim();
    const fsPercentage = getByTestId('final-score-percentage').textContent.trim()


    expect(correctAnswers).toBe('Correct: 2')
    expect(incorrectAnswers).toBe('Wrong: 3')
    expect(questionsAnswered).toBe('Total Answered: 5')
    expect(fsPercentage).toBe('Final Score: 40%')
  });

  test('it should dispatch an action to restart the quiz', () => {
    const { getByTestId } = renderWithContext(contextValue);
    const restartBtn = getByTestId('restart-btn');

    fireEvent.click(restartBtn)

    expect(contextValue.dispatch).toHaveBeenCalled();
  });

});