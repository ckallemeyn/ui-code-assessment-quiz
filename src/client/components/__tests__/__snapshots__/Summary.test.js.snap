// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Summary it should render 1`] = `
<div>
  <div
    class="ui container"
    data-testid="summary"
  >
    <div
      class="ui one column grid"
    >
      <div
        class="row"
      >
        <div
          class="column"
        >
          <h2>
            Summary
          </h2>
          <p>
            Correct: 
            <strong>
               
              0
               
            </strong>
          </p>
          <p>
            Incorrect: 
            <strong>
               
              0
               
            </strong>
          </p>
          <p>
            Total Answered: 
            <strong>
               
              0
               
            </strong>
          </p>
          <p>
            Final Score: 
            <strong>
               
              0
              % 
            </strong>
          </p>
          <button
            class="ui primary button"
            data-testid="restart-btn"
            type="text"
          >
            Restart Quiz
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
